In [1]:
import numpy as np
import librosa
import librosa.display
import os
import cv2
import matplotlib.pyplot as plt
import time

C:\ProgramData\Anaconda3\envs\tensorflow\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.3
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


**Please read UrbanSound8K_README.txt to better understand getting class names**

In [2]:
f_name = '7061-6-0-0.wav'
print(type(f_name))
a = f_name.split('-')
print(type(a[1]))

<class 'str'>
<class 'str'>


In [3]:
def get_class_name(file_name):
    """
        file_name: string
    return
        class_name: string
    """
    a = file_name.split('-')
    return a[1]

In [4]:
def create_dataset(audio_folder):
    """
        audio_folder: string
    return
        wav_data_array: a list of spectrograms
        class_name: a list of spectrograms' class names
    """
   
    wav_data_array=[]
    file_names = []
    class_name=[]
   
    for fold in os.listdir(audio_folder):
        for file_name in os.listdir(os.path.join(audio_folder, fold)):
            wav_path= os.path.join(audio_folder, fold,  file_name)
            data, fs = librosa.load(wav_path)
            spec = create_spectrogram(data)
            wav_data_array.append(spec)
            class_ = get_class_name(file_name)
            class_name.append(class_)
            file_names.append(file_name)
            
    return wav_data_array, file_names, class_name

In [5]:
def load_UrbanSound8K_dataset():
    """
    return
        datas, labels: numpy ndarrays
    """
    audio_folder_DIR = './UrbanSound8K/audio'
    begin = time.time()
    datas, names, labels = create_dataset(audio_folder_DIR)
    end = time.time()
#    datas = np.array(datas)
#    labels = np.array(labels)
    return datas, labels, names

In [6]:
def create_spectrogram(y):
    """
        y: a 1D numpy ndarray of .wav file
    
    return
        spec: a 2D numpy ndarray of desibels
    """
    
    spec = librosa.feature.melspectrogram(y=y)
    spec_conv = librosa.amplitude_to_db(spec, ref=np.max)
    return spec_conv

In [7]:
x, y, names = load_UrbanSound8K_dataset()

C:\ProgramData\Anaconda3\envs\tensorflow\lib\site-packages\librosa\util\decorators.py:88: UserWarning: n_fft=2048 is too small for input signal of length=1103
  return f(*args, **kwargs)
C:\ProgramData\Anaconda3\envs\tensorflow\lib\site-packages\librosa\util\decorators.py:88: UserWarning: n_fft=2048 is too small for input signal of length=1323
  return f(*args, **kwargs)
C:\ProgramData\Anaconda3\envs\tensorflow\lib\site-packages\librosa\util\decorators.py:88: UserWarning: n_fft=2048 is too small for input signal of length=1523
  return f(*args, **kwargs)


**audio dosyasının boyutu 6,6 GB o yüzden load işlemi 15-16 dakika sürüyor**

In [8]:
len(x)

8732

In [9]:
len(y)

8732

In [10]:
type(x[0])

numpy.ndarray

In [11]:
np.shape(x[0])

(128, 173)

In [12]:
def create_folder(file_name, m):
    for i in range(int(m)+1):
        try:
            os.makedirs(file_name + str(i))
        except OSError:
            print(file_name + str(i) +" already exist")
            pass

In [13]:
create_folder('./spec/', max(y))

In [14]:
def save_spectrograms(x, y, names, folder):
    """
        x: list of ndarrays
        y: list of strings
    """
    m = len(x)
    for i in range(m):
        tmp_list = names[i].split('.')
        tmp_list[1] = "png"
        name = '.'.join(tmp_list)
        path = os.path.join(folder, y[i], name)
        fig, ax = plt.subplots()
        img = librosa.display.specshow(x[i], ax=ax)
        fig.savefig(path, bbox_inches = 'tight', pad_inches = 0)
        plt.close()
        

In [15]:
save_spectrograms(x, y, names, './spec')

**8732 spectrogram olduğu için bu işlem de 15-16 dk sürüyor**